overlord/ifacestate/apparmorprompting/prompting.go

```diff
@@ -0,0 +1,475 @@
+	return p.tomb.Wait()
+}
+
+// GetPrompts returns all prompts for the user with the given user ID.
+func (p *Prompting) GetPrompts(userID uint32) ([]*requestprompts.Prompt, error) {
+	return p.prompts.Prompts(userID)
+}
+
+// GetPromptWithID returns the prompt with the given ID for the given user.
+func (p *Prompting) GetPromptWithID(userID uint32, promptID prompting.IDType) (*requestprompts.Prompt, error) {
+	return p.prompts.PromptWithID(userID, promptID)
+}
+
+// HandleReply checks that the given reply contents are valid, satisfies the
+// original request, and does not conflict with any existing rules (if lifespan
+// is not "single"). If all of these are true, sends a reply for the prompt with
+// the given ID, and both creates a new rule and checks any outstanding prompts
+// against it, if the lifespan is not "single".
+func (p *Prompting) HandleReply(userID uint32, promptID prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (satisfiedPromptIDs []prompting.IDType, retErr error) {
+	prompt, err := p.prompts.PromptWithID(userID, promptID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Outcome and lifesnap are validated while unmarshalling, and duration is
+	// validated when the rule is being added. So only need to validate
+	// constraints.
+	if err := constraints.ValidateForInterface(prompt.Interface); err != nil {
+		return nil, err
+	}
+
+	// Check that constraints matches original requested path.
+	// AppArmor is responsible for pre-vetting that all paths which appear
+	// in requests from the kernel are allowed by the appropriate
+	// interfaces, so we do not assert anything else particular about the
+	// constraints, such as check that the path pattern does not match
+	// any paths not granted by the interface.
+	// TODO: Should this be reconsidered?
+	matches, err := constraints.Match(prompt.Constraints.Path())
+	if err != nil {
+		return nil, err
+	}
+	if !matches {
+		return nil, fmt.Errorf("constraints in reply do not match original request: '%v' does not match '%v'; please try again", constraints, prompt.Constraints)
+	}
+
+	// TODO: once support for sending back bitmask of allowed permissions lands,
+	// do we want to allow only replying to a select subset of permissions, and
+	// auto-deny the rest?
+	contained := constraints.ContainPermissions(prompt.Constraints.RemainingPermissions())
+	if !contained {
+		return nil, fmt.Errorf("replied permissions do not include all requested permissions: requested %v, replied %v; please try again", prompt.Constraints.RemainingPermissions(), constraints.Permissions)
+	}
+
+	// TODO: a lock should be held while checking for conflicts with other rules
+	// so that if the rule is eventually removed due to an error, no prompts can
+	// have been matched against it in the meantime.
+	// A RWMutex over prompts and rules should work well, and could potentially
+	// replace the internal mutexes in those packages.
```

Comment:
can you explain a bit more? where is the conflict checking happening, inside AddRule?

Reply:
Yes, conflict checking is in `AddRule`, that's why we do that before replying, since we can't take back a reply once it's been sent to the kernel.

We definitely need to `AddRule` before `HandleNewRule`, so that there is no opportunity for a new prompt to be added between the checks against existing prompts and adding the rule to the DB, after which point new prompts are checked against them.

And before checking the new rule against all existing prompts (`HandleNewRule`), we want to make sure we're not going to reject the reply and remove the new rule and ask the client to try again. So we want to `Reply` before `HandleNewRule`.

So the problem is that there's a window of time between {when the rule is added and when the reply can be rejected} when a new prompt could be received and matched against the soon-to-be-removed new rule. The result is that there's a reply sent to the kernel because of a rule which shouldn't really exist.

This is fairly unlikely, and a rule handled because of a user's attempted (and failed) reply probably follows the user's intention, it's just not perfectly consistent, since the presence of the rule for that short window of time is an implementation detail.

Actually, any sort of interaction with rules during this window has the potential to be inconsistent. So I think it may make sense to put a `RWMutex` on the `InterfacesRequestsManager` which is locked for writing whenever the prompts or rules DB is being modified in a way which requires interaction between the two DBs. So handling a new rule, for example, either from a reply or by POSTing to the rules endpoint. And hold a read lock in other cases.